cpp/bench/matrix/select.cu

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "../../test/matrix/select.cuh"
+
+#include <common/benchmark.hpp>
+
+#include <raft/random/rng.cuh>
+#include <raft/sparse/detail/utils.h>
+#include <raft/util/cudart_utils.hpp>
+
+#include <raft/matrix/detail/select_radix.cuh>
```

Comment:
If these are needed to be tested individually, I think we should expose public APIs for them. in general, we should only be benchmarking/testing public APIs. 

(I know we are benchmarking the balanced kmeans which is currently in detail, but we're working to move this out into the public).

Reply:
The implementation variants in radix- and warpsort- files weren't and were never meant to be public. We've been automatically selecting the implementation in the public wrappers based on the runtime arguments. I see a few issues in making all of them public:

  1. There are a few rather experimental implementations (in warpsort- version), which are not meant to be used directly and probably may be deleted after we do extensive benchmarking.
  2. The api of all these functions is changing a little bit, it's going to be difficult to maintain this in the public namespace. We're also waiting for more implementations to arrive :) (faster radix-based selection)
  3. The api not very convenient for an end-user and contains template parameters that are for performance tweaking and implementation detail (block size for radix, warp-level implementation template for warpsort)
  4. All-in-all I don't see any benefit in imposing on the end user a burden of selecting one implementation of an algorithm among a dozen variants. I believe, we should continue to provide just one wrapper and be able to tweak the selection logic as new implementations and optimizations arrive to raft.

At the same time, we cannot remove the tests for individual variants, because the selection logic of the public wrapper makes it impossible to test full input ranges for every variant.
And we really need the individual benchmarks, so that we can compare the implementations and then adjust the selection logic (and repeat this as new implementations and optimizations arrive).

Reply:
> The implementation variants in radix- and warpsort- files weren't and were never meant to be public. 

I think this begs the questions as to how much we should expose and how advanced we should consider our users to be.  I'm not opposed to keeping these internal if we have a good well-defined (and well-documented) set of rules guiding the choice, resource, and runtime trade-offs of each. 

> At the same time, we cannot remove the tests for individual variants, because the selection logic of the public wrapper makes it impossible to test full input ranges for every variant.

I'm not saying we should remove the tests for individual variants but testing internals does not replace the need to test the public APIs comprehensively and exhaustively with all available options. After all, the users interact with the public APIs and the quality of their user experience should be better guaranteed by successfully passing tests.

The problem that arises from testing only internals and not also exhaustively testing the public APIs is that we gain a false sense of correctness when a developer makes changes to the public APIs.

Reply:
At the moment, we have [8 implementations/combinations](https://github.com/rapidsai/raft/blob/a31e61e72e66a4595c1954751c053476da2170c7/cpp/test/matrix/select_k.cuh#L44-L53). At least one more (new advances in radix-based select) in the works, and some of the current ones in `warp_sort` may be strictly worse than others. I hope we can keep all of them in detail, so that we can improve the implementation selection logic and remove some of the implementations in future.

Perhaps, it's not clear behind the templates, but the public API interface is the most tested in `test/matrix/select_k.cu`. Besides the [simple tests against static in-out values](https://github.com/rapidsai/raft/blob/a31e61e72e66a4595c1954751c053476da2170c7/cpp/test/matrix/select_k.cu#L327), we compare the outputs of [the various implementations against the public one](https://github.com/rapidsai/raft/blob/a31e61e72e66a4595c1954751c053476da2170c7/cpp/test/matrix/select_k.cu#L415-L441); this should ensure that the public-selected-implementation is compared against other non-selected implementations for various parameter ranges.
I think, this should cover the public interface well, but I'd be happy to add more tests if you have any further suggestions.